integration_test/third_party_apps_test.go

```diff
@@ -327,6 +328,110 @@ func runSingleTest(ctx context.Context, logger *logging.DirectoryLogger, vm *gce
 	return nonRetryable, nil
 }
 
+// Returns the authoritative set of all apps available for testing.
+// The authoritative list corresponds to the directory names under
+// integration_test/third_party_apps_data/applications
+func determineAllApps(t *testing.T) map[string]bool {
+	allApps := make(map[string]bool)
+
+	files, err := os.ReadDir("third_party_apps_data/applications")
+	if err != nil {
+		t.Fatalf("got error listing files under third_party_apps_data/applications: %v", err)
+	}
+	for _, file := range files {
+		if file.IsDir() {
+			allApps[file.Name()] = true
+		}
+	}
+	return allApps
+}
+
+func modifiedFiles(t *testing.T) []string {
+	cmd := exec.Command("git", "diff", "--name-only", "upstream/master")
+	out, err := cmd.Output()
+	if err != nil {
+		t.Fatalf("got error calling `git diff`: %v", err)
+	}
+
+	return strings.Split(string(out), "\n")
+}
+
+// Determine what apps are impacted by current code changes.
+// Extracts app names as follows:
+//   apps/<appname>.go
+//   integration_test/third_party_apps_data/<appname>/
+// Checks the extracted app names against the set of all known apps.
+// If any file modified is not specifically app related, appends an `all`
+// entry to the set of impacted apps, to be handled by the caller.
+func determineImpactedApps(mf []string, allApps map[string]bool) map[string]bool {
+	impactedApps := make(map[string]bool)
+	for _, f := range mf {
+		if strings.HasPrefix(f, "apps/") {
+
+			// File names: apps/<appname>.go
+			f := strings.TrimPrefix(f, "apps/")
+			f = strings.TrimSuffix(f, ".go")
+
+			if _, ok := allApps[f]; ok {
+				impactedApps[f] = true
+			} else {
+				impactedApps["all"] = true
+			}
+
+		} else if strings.HasPrefix(f, "integration_test/third_party_apps_data/applications/") {
+			// Folder names: integration_test/third_party_apps_data/applications/<app_name>
+			f := strings.TrimPrefix(f, "integration_test/third_party_apps_data/applications/")
+			f = strings.Split(f, "/")[0]
+			// The directories here are already authoritative, no
+			// need to check against list.
+			impactedApps[f] = true
+
+		} else {
+			// Any other modified file means we should test
+			// everything.
+			impactedApps["all"] = true
```

Comment:
I think this will result in "test everything" to occur all the time, since PRs that modify e.g. couchdb.go will also tend to modify various files outside of that directory. Examples:

https://github.com/GoogleCloudPlatform/ops-agent/pull/394/files
https://github.com/GoogleCloudPlatform/ops-agent/pull/379/files

I guess we might be able to ignore changes:
- under docs/
- under confgenerator/testdata/
- stuff like integration_test/third_party_apps_data/agent/ops-agent/linux/enable_hadoop, or better yet, rename this kind of file to integration_test/third_party_apps_data/applications/hadoop/enable

(hmm, not sure what to do about changes to integration_test/third_party_apps_data/agent/ops-agent/linux/supported_applications.txt, i guess ignore them and rely on the other heuristics instead)

It's tricky to set up the exclusions exactly right to avoid triggering "all" for PRs that touch an app... hmm

Reply:
Per our offline discussion, the latest revision always runs all apps for the default platform, and also runs all platforms for any app deemed to be "impacted".